### DISCUSSION TITLE
Fix Javascript numeric properties type detection

### DISCUSSION THREAD

**Discussion by riprwwa** [cefsharp/CefSharp#3853] [General]
When calling C# code from Javascript one can send objects as parameters.
These objects get translated to C# as `System.Dynamic.ExpandoObject` instances.
The types of these properties are inferred by the `SerializeV8Object` method in [`CefSharp.BrowserSubprocess.Core\Serialization\V8Serialization.cpp`](https://github.com/cefsharp/CefSharp/blob/cefsharp/94/CefSharp.BrowserSubprocess.Core/Serialization/V8Serialization.cpp#L43).

If setting a numeric property's value to something that fits into 32 bits, the property's type is inferred to be `System.Int32`.
If the value is greater than 2147483647 (which is 2^31-1) then the value is mistakenly interpreted to be a negative number:
- `2147483647` is correctly seen as `2147483647`
- `2147483648` is **in**correctly seen as `-2147483648`
- `2147483649` is **in**correctly seen as `-2147483647` and so on

One can easily test this by evaluating the following code in the CefSharp.Wpf.Example project, under "Evaluate Javascript (Async)"
`(() => { var a = { s: 2147483648 }; return a.s; })()`
and see that the result is a negative number.

@amaitland has suggested a workaround in a discussion in [Gitter](https://gitter.im/cefsharp/CefSharp):
```
[...]
else if (obj->IsInt())
{
    //CEF Doesn't differentiate between UINT and INT
    //so we have to do some additional bounds checking
    //To make sure the correct value is returned
    auto uIntVal = obj->GetUIntValue();
    if (uIntVal > INT_MAX)
    {
        //We have to use double as there is no IPC support
        //for UInt
        list->SetDouble(index, obj->GetDoubleValue());
    }
    else
    {
        list->SetInt(index, obj->GetIntValue());
    }
}
[...]
```
which does the trick in both the code I posted above, and my team's specific case.
Seeing as Javascript [always treats numbers as double](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number) anyway, is there a reason not to apply the suggested fix?

 I can create a pull request if need be, just wondering if/what I'm missing.

Thanks!

**Comment by amaitland** [comment]
Thanks for following up.

> Seeing as Javascript [always treats numbers as double](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number) anyway, is there a reason not to apply the suggested fix?

It's simply a matter of testing. An `xUnit` test will need to be added to validate the behaviour.
